### PR TITLE
Added unenroll program(s) feature

### PR DESCRIPTION
**Pull request by amir-qayyum-khan**
#### What are the relevant tickets?
fixes https://github.com/mitodl/micromasters/issues/4056

#### What's this PR do?
It adds a wizard to unenroll from one or many programs. And update UI on unenroll.

- [x] Added api
- [x] Add unenroll card on settings
- [x] Add unenroll programs wizard
  - [x] display enrolled programs
  - [x] Add multi select programs widget
  - [x] api call
  - [x] show spinner on api call
- [x] dashboard card when no program selected
- [x] Unit tests for front end
- [x] Unit tests for backend
 
#### How should this be manually tested?
- Enroll programs one or many
- Go to settings and click unenroll link
- then select program(s)
- check dashboard
- when no program left i.e you unenroll all, check new card on dashboard with enroll link

@pdpinch 
#### Screenshots (if appropriate)
-  screen shot 2018-08-03 at 5 25 22 pm
- screen shot 2018-08-02 at 3 07 14 pm
-  screen shot 2018-08-02 at 3 02 58 pm
-  screen shot 2018-08-02 at 3 08 47 pm
- screen shot 2018-08-03 at 4 30 50 pm

### CODE REVIEW

**Comment by pdpinch** [comment]
@roberthouse54 the pale blue background and shadow seems unusual. Is that what you had in mind?

**Comment by amir-qayyum-khan** [comment]
Removed shadow. It was not part of mockups
screen shot 2018-08-07 at 3 37 14 pm

**Comment by pdpinch** [comment]
Another question for @roberthouse54 -- Amir implemented an [error message when the user isn't logged into any programs](https://user-images.githubusercontent.com/10431250/43640777-a0d51964-973a-11e8-8ce9-c8fde7ded8f1.png). Does this work for you? Let us know if you'd like a review app to try it out.

**Comment by pdpinch** [comment]
@amir-qayyum-khan I'm surprised to see how much has been added to the dashboard code. I was thinking of this as being part of the settings UI. Is that all bundled together?

**Comment by amir-qayyum-khan** [comment]
> @amir-qayyum-khan I'm surprised to see how much has been added to the dashboard code. I was thinking of this as being part of the settings UI. Is that all bundled together?

@pdpinch yes this is all bundle.

File: static/js/components/LeaveProgramWizard.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+import _ from "lodash"
+
+import ProfileFormFields from "../util/ProfileFormFields"
+import type { Profile } from "../flow/profileTypes"
+import type {
+  AvailableProgramsState,
+  AvailableProgram
+} from "../flow/enrollmentTypes"
+import type { UIState } from "../reducers/ui"
+import { showDialog, hideDialog, setProgramsToUnEnroll } from "../actions/ui"
```
Comment (mbertrand) [inline static/js/components/LeaveProgramWizard.js]: Minor nitpick, can you move this import and the one below it above the type imports and below the `ProfileFormFields` import?
(https://github.com/mitodl/eslint-config#import-order)

**Comment by amir-qayyum-khan** [comment]
Done @mbertrand

**Comment by amir-qayyum-khan** [comment]
@mbertrand added more tests to fix coverage

File: static/js/components/LeaveProgramWizard.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+  props: {
+    dispatch: Dispatch,
+    profile: Profile,
+    programs: AvailableProgramsState,
+    ui: UIState
+  }
+
+  setDialogVisibility = (visible: boolean) =>
+    visible
+      ? showDialog(UNENROLL_PROGRAM_DIALOG)
```
Comment (rhysyngsun) [inline static/js/components/LeaveProgramWizard.js]: This is extra complexity that we don't need, you should just inline `showDialog` and `hideDialog`

File: static/js/components/LeaveProgramWizard.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        contentClassName="dialog unenroll-dialog"
+        className="unenroll-dialog-wrapper"
+        open={isVisible(dialogVisibility)}
+        onRequestClose={() => this.closeDialog()}
+        autoScrollBodyContent={true}
+        maxWidth="xs"
+        actions={dialogActions(
+          this.closeDialog,
+          this.unEnrollUserTask,
+          programsToUnEnrollInFlight,
```
Comment (rhysyngsun) [inline static/js/components/LeaveProgramWizard.js]: Should probably also disable if `programsToUnEnroll.length === 0`

File: static/js/components/LeaveProgramWizard.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  setDialogVisibility = (visible: boolean) =>
+    visible
+      ? showDialog(UNENROLL_PROGRAM_DIALOG)
+      : hideDialog(UNENROLL_PROGRAM_DIALOG)
+
+  enrolledInPrograms = () => {
+    const { programs } = this.props
+    return R.filter((program: AvailableProgram) => program.enrolled, programs)
```
Comment (rhysyngsun) [inline static/js/components/LeaveProgramWizard.js]: I think this would be better as a partial `const` outside the class and then call that in `renderUnenrollUI()`

File: static/js/actions/programs.js [hunk 5]
```diff
@@ -100,5 +103,42 @@ export const addProgramEnrollment = (
...
 
+export const UNENROLL_PROGRAM_DIALOG = "unenrollProgramDialog"
 export const CLEAR_ENROLLMENTS = "CLEAR_ENROLLMENTS"
 export const clearEnrollments = createAction(CLEAR_ENROLLMENTS)
+
+export const unEnrollProgramEnrollments = (
+  programIds: []
+): Dispatcher<?AvailableProgram> => {
+  return (dispatch: Dispatch) => {
+    dispatch(setUnEnrollApiInFlightState(true))
```
Comment (rhysyngsun) [inline static/js/actions/programs.js]: this should probably `await` here to avoid any potential race conditions

File: dashboard/views.py [hunk 6]
```diff
@@ -150,3 +152,37 @@ def post(self, request):
...
+    )
+    permission_classes = (permissions.IsAuthenticated, )
+
+    def post(self, request):
+        """
+        unenroll from MM program(s)
+        """
+        program_ids_serializer = UnEnrollProgramsSerializer(data=request.data)
+        program_ids_serializer.is_valid(raise_exception=True)
+        program_ids = program_ids_serializer.data['program_ids']
```
Comment (rhysyngsun) [inline dashboard/views.py]: This code should all get moved into the serializer, otherwise the serializer is just being used to parse a list of ints.

File: dashboard/views.py [hunk 7]
```diff
@@ -150,3 +152,37 @@ def post(self, request):
         return Response(
             data=enrollment.json
         )
+
+
+class UnEnrollPrograms(APIView):
+    """
+    api that unenroll user from one ore more programs
```
Comment (rhysyngsun) [inline dashboard/views.py]: Typo here

File: dashboard/views_test.py [hunk 8]
```diff
@@ -322,3 +322,106 @@ def test_enrollment(self, mock_refresh, mock_edx_enr, mock_index):  # pylint: di
...
+        cls.user = SocialUserFactory.create(social_auth__extra_data=social_extra_data)
+        cls.invalid_user = UserFactory.create()
+        cls.url = reverse('unenroll_programs')
+
+    def setUp(self):
+        super().setUp()
+        self.client.force_login(self.user)
+        self.program_enrollments = [ProgramEnrollmentFactory.create(user=self.user) for _ in range(2)]
+
+    def test_unenroll_all_programs(self):
```
Comment (rhysyngsun) [inline dashboard/views_test.py]: Should add a test for a mix of valid and invalid program ids

File: dashboard/views.py [hunk 9]
```diff
@@ -150,3 +152,37 @@ def post(self, request):
...
+            program_id__in=program_ids,
+            user=request.user
+        )
+        response = []
+        for program_enrollment in program_enrollments:
+            response.append({
+                'program_id': program_enrollment.program_id,
+                'title': program_enrollment.program.title
+            })
+            program_enrollment.delete()
```
Comment (rhysyngsun) [inline dashboard/views.py]: Are we confident in the side effects of outright deleting the `ProgramEnrollment` here? We generally don't delete data so we should verify that no parts of our application (particularly any scheduled tasks). Easy ones to find will be any `ProgramEnrollment.objects.get()` queries that aren't error-handled, but I suspect there may be more difficult edge cases to tease out. We may want to consider just marking these and inactive and filtering them out instead, although I realize that might be just as laborious.

Comment (pdpinch) [reply dashboard/views.py]: It's been my practice to delete ProgramEnrollment records when users wish to have their account deleted. I know this isn't exactly the same though, so I guess there are more cases to consider. 

If we take your suggestion of marking them inactive, will we need a migration, along with updating the code to filter out inactive ProgramEnrollments?

Comment (rhysyngsun) [reply dashboard/views.py]: Yes, well need both. I guess my point was, the amount of effort to do that is probably equal to or possibly greater than the amount of effort needed to shore up existing code. that doesn't handle situations where the record doesn't exist and is expected to, because those are probably the minority of all `ProgramEnrollment` code.

TLDR: deleting is **probably** still the best option (dragons aside), but we definitely need to test and/or audit at least the obvious code points that use these records to head off any errors. We probably won't catch them all, but unhandled `get(user=?, program=?)` calls are obvious candidates for where you'd hit that kind of error.

Comment (amir-qayyum-khan) [reply dashboard/views.py]: i tested all pages after removing all or single program, they are working fine.

**Comment by amir-qayyum-khan** [comment]
@pdpinch @rhysyngsun should we remove staff role if user as well, to reindex work properly.

File: dashboard/views.py (hunk 9 above)
Comment (amir-qayyum-khan) [reply dashboard/views.py]: @rhysyngsun any thoughts

Comment (rhysyngsun) [reply dashboard/views.py]: @amir-qayyum-khan you should test code that isn't on pages too, such as celery tasks, that query this.

**Comment by pdpinch** [comment]
@amir-qayyum-khan what's the status of this?

**Comment by amir-qayyum-khan** [comment]
celery tasks and signals were tested manually. I think i need to add new unit tests for them.

**Comment by amir-qayyum-khan** [comment]
Done @rhysyngsun

**Comment by amir-qayyum-khan** [comment]
Done @mbertrand,  I added tests related get(user=?, program=?) use in MM app and js test.
